hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+    int init(void);
+    bool scanning(void);
+    int setScanParams(const hal_ble_scan_params_t* params);
+    int getScanParams(hal_ble_scan_params_t* params) const;
+    int startScanning(on_ble_scan_result_cb_t callback, void* context);
+    int stopScanning(void);
+    void toPlatformScanParams(ble_gap_scan_params_t* params) const;
+
+private:
+    uint8_t isScanning_: 1;                                 /**< If it is scanning or not. */
+    hal_ble_scan_params_t scanParams_;                      /**< BLE scanning parameters. */
+    os_semaphore_t scanSemaphore_;                          /**< Semaphore to wait until the scan procedure completed. */
+    uint8_t scanReportBuff_[BLE_MAX_SCAN_REPORT_BUF_LEN];   /**< Buffer to hold the scanned report data. */
+    ble_data_t bleScanData_;                                /**< BLE scanned data. */
+    Vector<hal_ble_addr_t> cache_;                          /**< Cached address of scanned devices to filter-out duplicated result. */
+    Vector<hal_ble_gap_on_scan_result_evt_t> pendingResults_; /**< Caches the scanned advertising data until the scan response data is captured. */
+    BleObject* bleObj_;                                     /**< Pointer to the BLE Object. */
+    AtomicAllocedPool observerScannedDataPool_;             /**< Pool to allocate memory for scanned data. */
+    on_ble_scan_result_cb_t scanResultCallBack_;            /**< Callback function on scan result. */
+    void* context_;                                         /**< Context of the scan result callback function. */
+
+    static const uint16_t OBSERVER_SCANNED_DATA_POOL_SIZE = 1024;
+
+    bool chachedDevice(const hal_ble_addr_t& address) const;
+    hal_ble_gap_on_scan_result_evt_t* chachedResult(const hal_ble_addr_t& address);
+    void removeResult(const hal_ble_addr_t& address);
+    void clearResult(void);
+    int continueScanning(void);
+    static void observerEventProcessedHook(const hal_ble_evts_t *event, void* context);
+    static void processObserverEvents(const ble_evt_t* event, void* context);
 };
 
-struct Conn {
-    uint16_t handle;
-    uint16_t paramUpdateAttempts;
+class BleObject::ConnectionsManager {
+public:
+    typedef struct {
+        uint8_t role;
+        uint16_t connHandle;
+        hal_ble_conn_params_t effectiveConnParams;
+        hal_ble_addr_t peer;
+    } BleConnection_t;
+
+    ConnectionsManager(BleObject* obj);
+    ~ConnectionsManager() = default;
+    int init(void);
+    int setPPCP(const hal_ble_conn_params_t* ppcp);
+    int getPPCP(hal_ble_conn_params_t* ppcp) const;
+    bool connecting(void) const;
+    bool connected(const hal_ble_addr_t* address=nullptr);
+    int connect(const hal_ble_addr_t* address);
+    int connectCancel(void);
+    int disconnect(uint16_t connHandle);
+    int updateConnectionParams(uint16_t connHandle, const hal_ble_conn_params_t* params);
+    int getConnectionParams(uint16_t connHandle, hal_ble_conn_params_t* params);
```

Comment:
`const` ?

Reply:
In this function, it invokes a non const method `fetchConnection()`, which returns a writable pointer, when declaring this function as const, compiler will throw an error.